_drafts/test-automation-renode.md

```diff
@@ -0,0 +1,886 @@
+way to easily run tests on a device.
+
+A common approach is to use a CLI-based shell over serial, which we wrote about
+in our post [Building a Tiny CLI Shell for Tiny
+Firmware]({% post_url 2020-06-09-firmware-shell %}#integration--automated-tests).
+
+The idea is relatively simple.
+
+1. Create a test which will execute a series of shell commands on a device.
+2. After each command or series of commands, the results will be compared
+   against expected values.
+3. At the end of the test run, the harness determines whether the test was
+   successful or not.
+
+Thankfully for us, Renode has sever utilities and helpers built-in that we can
+take advantage of.
+
+## Setup
+
+### Example Repo
+
+All the code in this post was written for the STM32F429 MCU by ST Micro. While
+the examples run fine on the STM32F429i discovery board, they were developed in
+Renode, a popular MCU emulation platform.
+
+You can find the complete code example for this blog post in a separate
+[Github repository](https://github.com/memfault/interrupt-renode-test-automation).
+
+```
+$ git clone --recurse-submodules https://github.com/memfault/interrupt-renode-test-automation.git
+```
+
+### Toolchain
+
+I used the following tools to build my firmware:
+
+- GCC 8.3.1 / GNU Arm Embedded Toolchain as our compiler[^gnu_toolchain]
+- GNU Make 4.2.1 as the build system
+
+Running automated tests with Renode, unfortunately, requires Python 2.7. To keep
+things simple for me, I decided to use a Conda environment to keep this entire
+environment isolated.
+
+```
+$ conda create -n renode
+$ conda activate renode
+$ conda install -c conda-forge -c memfault \
+      python=2.7 gcc-arm-none-eabi make=4.2.1
+
+$ which arm-none-eabi-gcc
+/Users/tyler/miniconda3/envs/renode/bin/arm-none-eabi-gcc
+```
+
+Perfect, I now have `arm-none-eabi-gcc` and `python2.7` in my path.
+
+> Want to learn more about Conda? Check out my previous post on using [Conda > >
+> environments for embedded > >
+> development]({% post_url 2020-01-07-conda-developer-environments %}).
+
+It turns out the Renode team packages the application in a
+[Renode Conda package](https://anaconda.org/antmicro/renode), but at the time of
+writing this post, not all platforms had the latest version released (v1.9,
+released in March). For that reason, I chose to install it locally for this
+post.
+
+### Renode & Robot Framework
+
+To install Renode, I've found the best instructions to be on the
+[Renode Github Page](https://github.com/renode/renode#installation). In the
+example repo, I've created a `start.sh` script which points to my local Renode
+installation on my Mac. If you are using another platform, you'll have to update
+this script to point to your own Renode start script.
+
+```bash
+#!/bin/sh
+
+sh /Applications/Renode.app/Contents/MacOS/macos_run.command renode-config.resc
+```
+
+To run Robot Framework alongside Renode, we'll also need to clone the Renode
```

Comment:
First real time mentioning robot framework I believe. Might try to quickly mention what it is